fuel-vm/src/interpreter/blockchain.rs

```diff
@@ -487,39 +479,28 @@ impl<'vm, S, I> LoadContractCodeCtx<'vm, S, I> {
             return Err(PanicReason::ExpectedUnallocatedStack.into())
         }
 
-        let contract_id = a as usize;
-        let contract_id_end = checked_add_usize(ContractId::LEN, contract_id)?;
+        // Validate arguments
+        let contract_id = ContractId::from(read_bytes(self.memory, a)?);
         let contract_offset = b as usize;
         let length = bytes::padded_len_usize(c as usize);
+        let dst_range = MemoryRange::new(ssp, length)?;
 
-        let memory_offset = ssp as usize;
-        let memory_offset_end = checked_add_usize(memory_offset, length)?;
+        if dst_range.end >= *self.hp as usize {
+            // Would make stack and heap overlap
+            return Err(PanicReason::MemoryOverflow.into())
+        }
 
-        // Validate arguments
-        if memory_offset_end >= *self.hp as usize
-            || contract_id_end as Word > VM_MAX_RAM
-            || length > MEM_MAX_ACCESS_SIZE as usize
-            || length > self.contract_max_size as usize
-        {
+        if length > self.contract_max_size as usize {
             return Err(PanicReason::MemoryOverflow.into())
         }
 
         // Clear memory
-        self.memory[memory_offset..memory_offset_end].fill(0);
+        self.memory[dst_range.usizes()].fill(0);
 
-        // Fetch the contract id
-        let contract_id: &[u8; ContractId::LEN] = &self.memory
-            [contract_id..contract_id_end]
-            .try_into()
-            .expect("This can't fail, because we checked the bounds above.");
-
-        // Safety: Memory bounds are checked and consistent
-        let contract_id = ContractId::from_bytes_ref(contract_id);
-
-        self.input_contracts.check(contract_id)?;
+        self.input_contracts.check(&contract_id)?;
```

Comment:
maybe we should perform this validation before filling the memory with zeros? It gives us a way to bail out of execution earlier before performing a potentially large memset.